internal/scheduler/ingester.go

```diff
@@ -0,0 +1,777 @@
+			partitionIdx := msg.ID().PartitionIdx()
+			if srv.lastMessageIdByPartition == nil {
+				srv.lastMessageIdByPartition = make(map[int32]pulsar.MessageID)
+			}
+			if lastMessageId, ok := srv.lastMessageIdByPartition[partitionIdx]; ok {
+				msgIsOufOfOrder, err := pulsarutils.FromMessageId(lastMessageId).GreaterEqual(msg.ID())
+				if err != nil {
+					return err
+				}
+				if msgIsOufOfOrder {
+					// pulsarutils.PulsarMessageId prints nicely, so we convert to those.
+					return fmt.Errorf(
+						"unexpected seek detected: received messages out of order for topic %s: id of received message is %s, but the previous message has id %s",
+						msg.Topic(), pulsarutils.FromMessageId(msg.ID()), pulsarutils.FromMessageId(lastMessageId),
+					)
+				}
+			}
+			srv.lastMessageIdByPartition[partitionIdx] = msg.ID()
+
+			var err error
+			batches, err = srv.ProcessMessage(ctxWithLogger, msg, batches)
+			if err != nil {
+				logging.WithStacktrace(messageLogger, err).Warnf("processing message failed; ignoring")
+			}
+
+			// If we have more than 1 batch, apply all but the last immedtately.
+			batchesToApply := make([]*Batch, 0)
+			if len(batches) > 1 {
+				batchesToApply = append(batchesToApply, batches[:len(batches)-1]...)
+				batches = []*Batch{batches[len(batches)-1]}
+			}
+
+			// If the most recent batch is large enough,
+			// apply it immedtately to avoid batches becoming too large.
+			if len(batches) == 1 {
+				batch := batches[len(batches)-1]
+				if srv.ShouldWrite(batch) {
+					batchesToApply = append(batchesToApply, batch)
+					batches = make([]*Batch, 0)
+				}
+			}
+
+			// Batches to be applied are forwarded to a separate goroutine responsible for writing to postgres.
+			for _, batch := range batchesToApply {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case batchChannel <- batch:
+				}
+			}
+
+		case <-ticker.C: // Periodically send batches to be applied to postgres.
+			for _, batch := range batches {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case batchChannel <- batch:
+				}
+			}
+			batches = make([]*Batch, 0)
+		}
+	}
+}
+
+func (srv *Ingester) ProcessMessage(ctx context.Context, msg pulsar.Message, batches []*Batch) ([]*Batch, error) {
+	if msg == nil {
+		return batches, nil
+	}
+	log := ctxlogrus.Extract(ctx)
+
+	// Create a fresh batch if we don't already have one.
+	if len(batches) == 0 {
+		batches = append(batches, NewBatch())
+	}
+	batch := batches[len(batches)-1]
+
+	// Store the message id in the batch.
+	// So we can ack messages once they've been written to postgres.
+	//
+	// It's fine to write to postgres and then not ack, since writes are idempotent.
+	// But it's not fine to ack messages not written to postgres,
+	// since skipping messages isn't detected.
+	//
+	// We defer adding the message id to ensure it's added to the last batch created
+	// from this sequence.
+	defer func() {
```

Comment:
If I'm honest I'm not a fan of using a `defer` to modify the return value of the function after the return call.  I guess it's the principle of least surprise in that people don't expect their return object to be modified after the event.